lib/cloud/asset.js

```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
+    this.signerToken = null;
+  }
+
+  refreshSignerToken() {
+    const appName = this.settings.appName;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expire = Math.floor(Date.now() / 1000) + duration;
+    const token = settings.cloudAssetToken;
+    const host = this.settings.cloudAssetHost;
+    const url = `${host}/token/${appName}`;
+
+    return superagent.get(url)
+      .accept('application/json')
+      .set('Authorization', `Bearer ${token}`)
+      .query({expired_at: expire.toString()})
+      .then(response => {
+        const body = response.body;
+        console.log('fuck', body);
+        this.signerToken = {
+          value: body.value,
+          expiredAt: Date.parse(body.expired_at),
+          extra: body.extra
+        };
+      });
+  }
+
+  needRefreshSignerToken() {
+    if (this.signerToken === null) {
+      return true;
+    }
+    if (this.signerToken.expiredAt < new Date()) {
+      return true;
+    }
+    return false;
+  }
+
+  sign(name) {
+    if (this.needRefreshSignerToken()) {
+      return this.refreshSignerToken()
+        .then(()=> this.sign(name));
+    }
+    const isPublic = !!this.settings.cloudAssetStorePublic;
+    const prefix = isPublic ?
+      this.settings.cloudAssetPublicPrefix :
+      this.settings.cloudAssetPrivatePrefix;
+    const appName = this.settings.appName;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expired = Math.floor(Date.now() / 1000) + duration;
+
+    const secret = this.signerToken.value;
+    const hash = crypto.createHmac('sha256', secret)
+      .update(appName)
+      .update(name)
+      .update(expired.toString())
+      .update(this.signerToken.extra)
+      .digest('base64');
+    const signature_and_extra =
+      encodeURIComponent(`${hash}.${this.signerToken.extra}`);
+
+    return `${prefix}/${appName}/${name}` +
+      `?expired_at=${expired}&signature=${signature_and_extra}`;
+  }
+}
+
+class S3Signer extends Signer {
+  constructor(_settings) {
+    super();
+    this.settings = _settings;
+    const key = _settings.assetStoreS3AccessKey;
+    const secret = _settings.assetStoreS3SecretKey;
+
+    let host;
+    let bucket;
+    const prefix = _settings.assetStoreS3URLPrefix;
+    if (prefix) {
+      const parsed = URL.parse(prefix);
+      host = parsed.host;
+      bucket = parsed.pathname;
+    } else {
+      const region = _settings.assetStoreS3Region;
+      host = `s3-${region}.amazonaws.com`;
+      bucket = _settings.assetStoreS3Bucket;
+    }
+
+    this.signer = S3URLSigner.urlSigner(key, secret, {
+      host: host
+    });
+    this.bucket = bucket;
+  }
+  sign(name) {
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const url = this.signer.getUrl('GET', name, this.bucket, duration);
+    return Promise.resolve(url);
+  }
+}
+/**
+ * Return a signer for the current configuration.
+ *
+ * @return {Signer}
+ */
+export function getSigner() {
```

Comment:
We only need one instance of signer. The name `getSigner` also imply this to me.

So I have two quesiton:
- What is the consideration on returning new instance on every `getSigner`?
- The name `signerFactory` will tell everyone that the new signer will return on every call.